nodejs/device/src/index.ts

```diff
@@ -175,11 +178,38 @@ class DeviceApp {
         );
       };
       this.twin.on("properties.desired.thief.pairing", listener);
-      const timer = setTimeout(
+      const pairingAttemptTimer = setTimeout(
         on_timeout,
         1000 * settings.pairingRequestSendIntervalInSeconds
       );
-    });
+
+      const finish = () => {
+        logger.warn("Pairing canceled.");
+        this.twin.removeListener("properties.desired.thief.pairing", listener);
+        clearTimeout(pairingAttemptTimer);
+        reject(new Canceled());
+      };
+
+      cancel = () => {
+        if (finished) {
+          return;
+        }
+        finish();
+      };
```

Comment:
You have 2 functions named `cancel` in here, one in the outer scope that sets finished=True and this one.  

If I'm reading this correctly, you return the first one to the caller of `pairingAttempt` and this one (and  `finish` immediately above) get created in local scope, aren't used, and then eventually get collected.  Is this right? 

Or, since this one doesn't have a `let` or `var` keyword, that implies that you're replacing the `cancel` variable in the outer scope.  But, if you've already returned the old value of `cancel`, replacing it doesn't benefit the user, does it? Or are there some execution rules that mean that the outer scope `cancel` already has the new value before `pairingAttempt` returns?

Closures and Promises sometimes hurt my brain, especially since each language has different rules about how they work.

Reply:
This was an attempt to handle the case where cancel is called before the executor of the promise is run. However after some testing I think the executor gets called synchronously when the Promise constructor is called, so this might not actually be needed. 

Objects are references, we can replace the `cancel` property and it will be changed outside the function. 
Consider this example, which prints "hello".
```js
function test() {
    let obj = {func: () => {console.log("hey")}};
    process.nextTick(() => {obj.func = () => {console.log("hello")}});
    return obj;
}

let obj = test();
process.nextTick(() => {obj.func()});
```

Reply:
Okay, so it's that the executor of the promise is called synchronously when the Promise constructor is called. This means that the assignment of `cancel` in the Promise executor happens before the `return`.